### DISCUSSION TITLE
Changes to lighting components

### DISCUSSION THREAD

**Discussion by Ivanzyfer** [FakeFishGames/Barotrauma#14881] [Ideas and suggestions]
Now the lighting components strongly illuminate the area around them and look not like an indication, but like poorly working lamps. It would be possible either to change their design to make them look attractive or to add an object "indicator light" that could be used as an indication of different parameters
![baro](https://github.com/user-attachments/assets/7d39301b-6ecd-4fc1-90b0-9f6e4a941a7b)

**Comment by Ek-Videogames** [comment]
Not to mention the light XML 'components' - even if they aren't even doing anything - still contribute to the game's processing overhead due to quirk with the light drawing engine.

As suggested, a custom component for indicator light - possibly one that uses a more optimized method of rendering?

( Also it'd be nice if regular lights with their color & alpha set to 0 didn't render either, as turning the 'built in' lights on various items like engines or large pumps disables them visually but still affects graphics performance & item 'component' updates. )

**Reply by Regalis11** [reply]
> due to quirk with the light drawing engine.

Could you elaborate on this? What quirk? 

As far as I can see, the only processing overhead these sorts of lights add is rendering the additive light on the light map, and I can't really think of how that could be done in a more optimized way.

**Reply by Ek-Videogames** [reply]
- Unfortunately this was years ago so I forget which dev said it specifically, but it was stated that even lights with 0 transparency were contributing another "layer" of additive rendering pass to the "lightmap"/shadows layer ( even though the in-editor UI doesn't count them as shadow casting lights, it still adds another 'light' to the total drawing ). The item would still have the lighting 'quad' drawn around it when viewed in debug mode. This was also way before the more recent optimization pass through the lighting & new culling system though so take it with a grain of salt.

It would lose the glowing aspect, but perhaps just an indicator which allows changing it's sprite color (similar to text display's set_color but it just changes the item color) is a possible workaround.

Also genuine question, since some lights are actually made up of two additive passes - one being the sphere of light being emitted, and the other being the additive 'mask' overlaid on the item itself to make it look like it's glowing (also this effect seems to respect sprite-depth unlike the 'global' lighting layer) - how possible is it to just make a light with a single pass for it's own "self-illumination"? So at least things like indicator lights which don't really need to cast light in the first place have their performance cost essentially halved.

![image](https://github.com/user-attachments/assets/38051d73-3085-4713-90c2-59f4f7914551)

![image](https://github.com/user-attachments/assets/581d3893-e333-47e4-8d92-6299f1139172)

**Reply by Regalis11** [reply]
That explanation is correct, the light does need to be drawn into the additive "light map" for it to glow. This has a negligible impact on performance though - it's just a single sprite that's being rendered, with no other fancy logic, no more expensive than let's say drawing one particle sprite (which there can be thousands of on the screen). I would not consider this a quirk: if we want a light to glow like this, we need to draw an additive sprite on it.

> Also genuine question, since some lights are actually made up of two additive passes - one being the sphere of light being emitted, and the other being the additive 'mask' overlaid on the item itself to make it look like it's glowing (also this effect seems to respect sprite-depth unlike the 'global' lighting layer) - how possible is it to just make a light with a single pass for it's own "self-illumination"? So at least things like indicator lights which don't really need to cast light in the first place have their performance cost essentially halved.

This is entirely possible - in the XML it can be done for example by setting the alpha of the light's color to 0, and setting the alpha of the additive sprite above 0 (which is what many vanilla items do). But in any case, the performance cost of these lights that don't cast shadows (calculating the shadows being the expensive part of the light rendering) is so negligible it wouldn't make much difference in practice.

**Comment by Ivanzyfer** [comment]
I was referring specifically to the appearance, the fact that the areas highlighted around do not look very nice

**Reply by Regalis11** [reply]
Yeah, the conversation above got kind of side-tracked. :D I agree these don't look that nice: they can be made to look better by making the range of the light a lot shorter, which I suppose we could do in vanilla subs (and maybe make the components default to a very short range too).
